torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -511,6 +511,125 @@ std::vector<int64_t> ComputeShapeFromReshape(
+  if (opset_version >= 14 && n->hasAttributeS("allowzero")) {
+    allowzero = n->i(attr::allowzero);
+    if (allowzero == 1 && reshape_has_zero) {
+      return reshape;
+    }
+  }
+
+  auto input_shape_size = static_cast<int>(input_shape_vector.size());
+  auto is_minus_one = [](c10::ShapeSymbol& ss) { return ss.value() == -1; };
+  auto it_minus_one =
+      std::find_if(reshape_vector.begin(), reshape_vector.end(), is_minus_one);
+  int minus_one_pos = it_minus_one == reshape_vector.end()
+      ? -1
+      // NOLINTNEXTLINE(bugprone-narrowing-conversions,cppcoreguidelines-narrowing-conversions)
+      : std::distance(reshape_vector.begin(), it_minus_one);
+
+  if (!reshape_has_zero && minus_one_pos == -1) {
+    return reshape;
+  }
+  std::vector<c10::ShapeSymbol> final_shape;
+  double shape_ratio = 1.0;
+  // sym_map is used to match shape symbols between the input and reshape.
+  // If there is a mismatch, the output shape cannot be estimated.
+  std::unordered_map<int64_t, int64_t> sym_map;
+  for (const auto i : c10::irange(input_shape_size)) {
+    if (input_shape_vector[i].is_static()) {
+      shape_ratio *= static_cast<double>(input_shape_vector[i].static_size());
+    } else {
+      auto value = input_shape_vector[i].value();
+      if (sym_map.find(value) != sym_map.end()) {
+        sym_map[value]++;
+      } else {
+        sym_map[value] = 1;
+      }
+    }
+  }
+  for (const auto i : c10::irange(reshape_size)) {
+    if (i != minus_one_pos) {
+      if (reshape_vector[i].value() != 0) {
+        if (reshape_vector[i].is_static()) {
+          shape_ratio /= static_cast<double>(reshape_vector[i].static_size());
+        } else {
+          auto value = reshape_vector[i].value();
+          if (sym_map.find(value) != sym_map.end()) {
+            sym_map[value]--;
+            if (sym_map[value] == 0) {
+              sym_map.erase(value);
+            }
+          } else {
+            return c10::nullopt;
+          }
+        }
+      } else {
+        if (input_shape_vector[i].is_static()) {
+          shape_ratio /=
+              static_cast<double>(input_shape_vector[i].static_size());
+        } else {
+          auto value = input_shape_vector[i].value();
+          if (sym_map.find(value) != sym_map.end()) {
+            sym_map[value]--;
+            if (sym_map[value] == 0) {
+              sym_map.erase(value);
+            }
+          } else {
+            return c10::nullopt;
+          }
+        }
+      }
+    }
+  }
+
+  if (!sym_map.empty()) {
+    return c10::nullopt;
+  }
+
+  for (const auto i : c10::irange(minus_one_pos)) {
+    c10::ShapeSymbol cur_shape(
+        reshape_vector[i].value() == 0 ? input_shape_vector[i]
+                                       : reshape_vector[i]);
+    final_shape.push_back(cur_shape);
+  }
+  final_shape.push_back(ShapeSymbol::fromStaticSize(
+      static_cast<int64_t>(std::round(shape_ratio))));
+  for (auto i = minus_one_pos + 1; i < reshape_size; i++) {
```

Comment:
I think this is wrong if `minus_one_pos == -1`, and it's not obvious to me why that can't happen.

Reply:
There is an early return logic
```python
  if (!shape_has_zero && minus_one_pos == -1) {
    return shape;
  }
```
Another
```python
  if (allowzero == 1 && shape_has_zero) {
    return shape;
  }
```
So the only case coming here when `minus_one_pos == -1` is that 
`shape_has_zero = True` and also `allow_zero == 0` 
I don't see pytorch can use 0 in the torch.shape, so it seems this condition is not satisfied, meaning that here `minus_one_pos` can never be -1.

Reply:
Can you please add a `TORCH_ASSERT` just in case the assumption is wrong (or becomes wrong in the future)?

Reply:
added